### PR TITLE
feat: add xml add-on

### PR DESCRIPTION
**Pull request by warlof**
## Proposed changes

I'm using Weblate to maintain XLIFF translation files from my Business Central packages.
Compiler is using non self-closing tags format when no data is stored in a node - and as a result, huge diff are currently generated after a unit is translated from Weblate.

The purpose of this changes is to provide an optional way to handle the output formatting of XLIFF files.

It relays on the following pull request which add a flag to the XLIFF storage https://github.com/translate/translate/pull/4751

## Checklist

- [x] Lint and unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added documentation to describe my feature.
- [x] I have squashed my commits into logic units.
- [x] I have described the changes in the commit messages.

### CODE REVIEW

File: weblate/addons/xliff.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class XLIFFCustomizeAddon(StoreBaseAddon):
+    name = "weblate.xliff.customize"
+    verbose = _("Customize XLIFF output")
+    description = _(
+        "Allows adjusting XLIFF output behavior, for example indentation or sorting."
+    )
+    settings_form = XLIFFCustomizeForm
+    multiple = True
```
Comment (nijel) [inline weblate/addons/xliff.py:34]: There is no reason to install such add-ons multiple times.

```suggestion
```

File: weblate/addons/xliff.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    icon = "language.svg"
+
+    def store_post_load(self, translation, store):
+        config = self.instance.configuration
+        tags_format = config.get("tags_format", "closing_tags")
+        if tags == "closing_tags":
+            store.XMLSelfClosingTags = False
```
Comment (nijel) [inline weblate/addons/xliff.py:41]: It should always set the attribute:

```suggestion
        store.XMLSelfClosingTags = config.get("tags_format", "closing_tags") == "self_closing_tags"
```

File: weblate/addons/xliff.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+from django.utils.translation import gettext_lazy as _
+
+from weblate.addons.base import StoreBaseAddon
+from weblate.addons.forms import XLIFFCustomizeForm
+
+
+class XLIFFCustomizeAddon(StoreBaseAddon):
+    name = "weblate.xliff.customize"
+    verbose = _("Customize XLIFF output")
```
Comment (nijel) [inline weblate/addons/xliff.py:29]: Maybe make it XML generic? With https://github.com/translate/translate/pull/4751, the feature is added to all Lisa subclasses.

**Review by nijel — CHANGES_REQUESTED** [review]
Besides other comments, the add-on is missing compatibility check. Either add `compat` attribute with supported file formats, or implement `can_install` method to check whether component file format is subclass of Lisa.

File: weblate/addons/xliff.py (hunk 3 above)
Comment (warlof) [reply weblate/addons/xliff.py:29]: Make sense 👍

**Comment by warlof** [comment]
@nijel I made some change localy, especially regarding the can_install override in order to check if yes or no we can use the formatter addon.

However, I'm still stuck for testing since when I go on the component addons page, I'm getting an exception stating the module does not exist.
I can't figure what I did wrong :(

```
ImproperlyConfigured at /addons/business-central/email-current-user-connector/
Error importing module weblate.addons.xml in WEBLATE_ADDONS: "No module named 'weblate.addons.xml'"
```

Also, in the `can_install` definition - does `cls.instance` refer to the Storage class (aka LISAfile) ?

**Comment by warlof** [comment]
hey, I'm back and it's working :')

it took me a while, but I figure what was going wrong.
currently, the quick dev using docker is binding volume against python 3.10 - however, my container were boot using python 3.11
I don't know if it's normal or not - but at least, once volume is bound to proper target path, it's going better :D
https://github.com/WeblateOrg/weblate/blob/c1874814d7ea30020fff92bb11a47af8f288765c/dev-docker/docker-compose.yml#L30

let's back to our main topic :

I had override the `can_install` method in order to determine dynamically if yes or no the formatter addon can be installed.
Personal notice : component -> format -> store

I also had fix an issue within the `store_post_load` method which is taking store as parameter but it appears to be the FileFormat instead.

And now, I can confirm it's working as expected 👍

File: weblate/formats/ttkit.py [hunk 4]
```diff
@@ -506,7 +506,7 @@ def all_store_units(self):
         )
 
     def delete_unit(self, ttkit_unit) -> Optional[str]:
-        self.store.removeunit(ttkit_unit)
+        return self.store.removeunit(ttkit_unit)
```
Comment (nijel) [inline weblate/formats/ttkit.py:509]: This is wrong, this method intentionally returns None. The problem with that is that adding `return None` is pointless :-). Anyway this is not related to topic of this PR.

File: weblate/addons/forms.py [hunk 5]
```diff
@@ -201,6 +201,20 @@ class JSONCustomizeForm(BaseAddonForm):
...
+
+    tags_format = forms.ChoiceField(
+        label=_("XML closing tags format"),
+        choices=[
+            ("self_closing_tags", _("Self-Closing Tags")),
+            ("closing_tags", _("Closing Tags")),
+        ],
+        required=True,
+        initial="self_closing_tags",
+    )
```
Comment (nijel) [inline weblate/addons/forms.py:215]: This needs rephrasing so that it is easier to understand. Also, making it checkbox might be nicer.

How about this? (It would need changes in the add-on as well)

```suggestion
    closing_tags = forms.Boolean(
        label=_("Include closing tag for blank XML tags"),
        required=False,
        initial=True,
    )
```

File: weblate/addons/xml.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        # component are attached to a file format which is defined by a loader
+        # we want to provide this package only for component using LISAfile as loader
+        return issubclass(component.file_format_cls.loader, LISAfile)
+
+    def store_post_load(self, translation, store):
+        """Hook triggered once component formatter has been loaded."""
+        config = self.instance.configuration
+        store.store.XMLSelfClosingTags = (
+            config.get("tags_format", "closing_tags") == "self_closing_tags"
+        )
```
Comment (nijel) [inline weblate/addons/xml.py:50]: Change to work with `BooleanField`:

```suggestion
        store.store.XMLSelfClosingTags = config.get("closing_tags", True)
```

**Review by nijel — COMMENTED** [review]
This needs tests as well. Once https://github.com/translate/translate/pull/4751 is merged and released, we can adjust the dependency to make sure the functionality is supported in the library.

File: weblate/addons/forms.py (hunk 5 above)
Comment (warlof) [reply weblate/addons/forms.py:215]: Definitely

**Comment by nijel** [comment]
Weblate now requires translate-toolkit 4.8.0 with needed changes. Can you please look into adding a testcase for this?

**Comment by warlof** [comment]
> Weblate now requires translate-toolkit 4.8.0 with needed changes. Can you please look into adding a testcase for this?

I will have to figure why I cannot run existing ones first.

What to be test ?
- if unchecked => self-closing tags
- if checked => non self-closing tags

**Comment by warlof** [comment]
I also had realign the branch against master `HEAD` :)

**Comment by warlof** [comment]
Hi there,

Is there hope it's being shipped with next 4.15 upcoming version ? That would be a great Xmas present 😇

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Comment by nijel** [comment]
I've updated the docs in f940c51b02
